spec/main_io_spec.rb

```diff
@@ -0,0 +1,81 @@
+require 'spec_helper'
+require 'main_io'
+
+class FakeStdin
+  @@string
+
+  def self.string
+    @@string
+  end
+
+  def self.string=(string)
+    @@string = string
+  end
+
+  def self.gets 
+    @@string
+  end
+end
+
+class FakeStdout
+  def self.puts(*output)
+    output.join("\n")
+  end
+end
+
+class FakeSpy
+  attr_accessor :times_called
+
+  def initialize
+    @times_called = 0
+  end
+
+  def watch(method:)
+    define_singleton_method(method) { |*| self.times_called += 1 }
+  end
+end
+
+describe MainIO do
+  context 'when accepting user input' do
+    it 'returns captured string' do
+      test_input = '1'
+      io = MainIO.new(stdin: FakeStdin, stdout: FakeStdout)
+      FakeStdin.string= '1'
+      input_result = io.gets
+      expect(input_result).to eql test_input
+    end
+
+    it 'uses stdin to capture the string' do
```

Comment:
If `MainIO` weren't calling `gets` on stdin, would the test before this one pass? I'm thinking it would not. I perceive the test before this one as using the fake stdin to avoid having to test the internal implementation (as opposed to the public behavior) of `MainIO` (namely that it calls `gets`). The fake does have to be implemented with knowledge of the interface of stdin, but the nice thing is that that knowledge only need exist in one place. 

On the other hand, I perceive this test here as being an example of testing the internal implementation of `MainIO` - it's testing more the private "how it gets done" than the public "what gets done" of `MainIO`. Testing implementation rather than behavior tends to make refactoring and changes more difficult, and offers less benefit as a check of correct behavior.

Sandy Metz (author of POODR) has a good talk that touches on this: https://www.youtube.com/watch?v=URSWYvyc42M

Reply:
Thanks for the video, suggestion and talking about it in standup. It makes a lot more sense now. I implemented the change, by removing the tests with spies since it only tested implementation. I believe the tests above it satisfies what's needed of IO's functionality.